File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import sys


def init(node, start, end) :
    if start == end :
        tree[node] = l[start]
    else :
        mid = (start + end) // 2
        tree[node] = init(node * 2, start, mid) + init(node * 2 + 1, mid + 1, end)
    return tree[node]


def subSum(node, start, end, left, right) :
    if start > right or end < left :
        return 0

    if left <= start and end <= right :
        return tree[node]

    mid = (start + end) // 2
    return subSum(node * 2, start, mid, left, right) + subSum(node * 2 + 1, mid + 1, end, left, right)


def update(node, start, end, index, diff) :
    if start > index or end < index :
        return

    tree[node] += diff

    if start != end :
        mid = (start + end) // 2
        update(node * 2, start, mid, index, diff)
        update(node * 2 + 1, mid + 1, end, index, diff)


N, M, K = map(int, (sys.stdin.readline()).split())
l = []
for _ in range(N) :
    l.append(int(sys.stdin.readline().strip()))

tree = [0] * (N *10)
init(1, 0, N - 1)

for _ in range(M + K) :
    a, b, c = map(int, (sys.stdin.readline()).split())
    if a == 1 :
        diff = c - l[b - 1]
        l[b - 1] = c
        update(1, 0, N - 1, b - 1, diff)
    else :
        print(subSum(1, 0, N - 1, b - 1, c - 1))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.*;
import java.util.*;

public class Main {

    static long[] elements, tree;

    public static long init(int start, int end, int node) {
        if (start == end) {
            return tree[node] = elements[start];
        }
        int mid = (start + end) / 2;
        return tree[node] = init(start, mid, node * 2) + init(mid + 1, end, node * 2 + 1);
    }

    public static long subSum(int start, int end, int left, int right, int node) {
        if (start > right || end < left) {
            return 0;
        }

        if (left <= start && end <= right) {
            return tree[node];
        }

        int mid = (start + end) / 2;
        return subSum(start, mid, left, right, node * 2) + subSum(mid + 1, end, left, right, node * 2 + 1);
    }

    public static void update(int start, int end, long diff, int index, int node) {
        if (start > index || index > end) {
            return;
        }
        tree[node] += diff;
        if (start != end) {
            int mid = (start + end) / 2;
            update(start, mid, diff, index, node * 2);
            update(mid + 1, end, diff, index, node * 2 + 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());

        int N = Integer.parseInt(tokenizer.nextToken());
        int M = Integer.parseInt(tokenizer.nextToken());
        int K = Integer.parseInt(tokenizer.nextToken());

        elements = new long[N + 1];
        for (int i=1;i < N + 1;i ++) {
            elements[i] = Long.parseLong(reader.readLine());
        }
        tree = new long[N * 4];
        init(1, N, 1);

        StringBuilder sb = new StringBuilder();
        for (int i=0;i < M+K;i ++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int a = Integer.parseInt(tokenizer.nextToken());
            int b = Integer.parseInt(tokenizer.nextToken());
            long c = Long.parseLong(tokenizer.nextToken());

            if (a == 1) {
                long diff = c - elements[b];
                elements[b] = c;
                update(1, N, diff, b, 1);
            }
            else {
                sb.append(subSum(1, N, b, (int) c, 1) + "\n");
            }
        }

        bw.write(sb.toString());
        bw.flush();
        bw.close();
        reader.close();
    }
}
